File: bindings/swift/QbtClient.swift
import Foundation

struct QbtClient {
    let baseURL: URL
    let token: String?

    init(baseURL: String = "http://127.0.0.1:8766", token: String? = nil) throws {
        guard let url = URL(string: baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/"))) else {
            throw URLError(.badURL)
        }
        self.baseURL = url
        self.token = token
    }

    func health() async throws -> [String: Any] {
        try await request(path: "/health", method: "GET", body: nil)
    }

    func sample(provider: String = "simulator", shots: Int = 1024, seed: Int = 42) async throws -> [String: Any] {
        try await request(
            path: "/v1/sample",
            method: "POST",
            body: ["provider": provider, "shots": shots, "seed": seed]
        )
    }

    func normalize(_ payload: [String: Any]) async throws -> [String: Any] {
        try await request(path: "/v1/normalize", method: "POST", body: payload)
    }

    private func request(path: String, method: String, body: [String: Any]?) async throws -> [String: Any] {
        guard let url = URL(string: path, relativeTo: baseURL) else {
            throw URLError(.badURL)
        }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw NSError(domain: "QBT", code: 1, userInfo: [NSLocalizedDescriptionKey: text])
        }
        guard let value = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw NSError(domain: "QBT", code: 2, userInfo: [NSLocalizedDescriptionKey: "non-object JSON"])
        }
        return value
    }
}

File: bindings/javascript/qbt-client.mjs
export class QbtClient {
  constructor(baseUrl = "http://127.0.0.1:8766", token = "") {
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.token = token;
  }

  async request(path, options = {}) {
    const headers = new Headers(options.headers || {});
    headers.set("Accept", "application/json");
    if (options.body) headers.set("Content-Type", "application/json");
    if (this.token) headers.set("Authorization", `Bearer ${this.token}`);
    const response = await fetch(`${this.baseUrl}${path}`, { ...options, headers });
    const text = await response.text();
    let payload;
    try {
      payload = JSON.parse(text);
    } catch {
      throw new Error(`QBT returned non-JSON response: ${text}`);
    }
    if (!response.ok) {
      throw new Error(payload.error || `QBT HTTP ${response.status}`);
    }
    return payload;
  }

  health() {
    return this.request("/health");
  }

  status(provider = "simulator", seed = 42) {
    return this.request(
      `/v1/status?provider=${encodeURIComponent(provider)}&seed=${encodeURIComponent(seed)}`,
    );
  }

  sample({ provider = "simulator", shots = 1024, seed = 42 } = {}) {
    return this.request("/v1/sample", {
      method: "POST",
      body: JSON.stringify({ provider, shots, seed }),
    });
  }

  normalize(payload) {
    return this.request("/v1/normalize", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  }
}
